### PR TITLE
Add instrumentation and clj-kondo support for ClojureScript

### PR DESCRIPTION
**Pull request by dvingo**
Addresses https://github.com/metosin/malli/issues/482 

Implements function instrumentation for ClojureScript code, attempting
to match as closely as possible to the Clojure implementation.

Adds separate namespaces to implement instrumentation support in cljs
because macros are required to dynamically change vars in cljs
(symbol literals are required at runtime).

malli.dev.cljs/start! does not feature a watch on the function schemas
atom because the var names need to be known at compile time in order
for a macro to produce valid cljs code. In cljs, hot code reloading
is the norm so this isn'ta major deficiency.

### CODE REVIEW

File: src/malli/clj_kondo.cljc [hunk 1]
```diff
@@ -175,15 +176,18 @@
...
       (assoc-in
-        acc [:linters :type-mismatch :namespaces (symbol (str ns)) name :arities arity]
+        acc [:linters :type-mismatch :namespaces ns name :arities arity]
         (select-keys data [:args :ret :min-arity])))
     {:linters {:unresolved-symbol {:exclude ['(malli.core/=>)]}}} xs))
 
 #?(:clj
    (defn emit! [] (-> (collect) (linter-config) (save!)) nil))
+
+#?(:clj
```
Comment (ikitommi) [inline src/malli/clj_kondo.cljc:192]: why is this needed?

File: src/malli/dev/cljs.clj [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+(defmacro start!
+  "Collects defn schemas from all loaded namespaces and starts instrumentation for
+   a filtered set of function Vars (e.g. `defn`s). See [[malli.core/-instrument]] for possible options.
+   Also emits clj-kondo type annotations.
+   This does NOT re-instrument functions if the function schemas change - use hot reloading to get a similar effect."
+  [options]
+  `(do
+     ~(mi/unstrument* nil)
+     ~(do (clj-kondo/save! {}) nil)
```
Comment (ikitommi) [inline src/malli/dev/cljs.clj:21]: why ensure a`nil` return here?

**Review by ikitommi — APPROVED** [review]
Looks great! Don't fully understand what happens at `malli.instrument.cljs`, but big thanks for making this!

File: src/malli/dev/cljs.clj (hunk 2 above)
Comment (dvingo) [reply src/malli/dev/cljs.clj:21]: No reason in particular - I can just have it return whatever `(clj-kondo/save! {})` returns.

File: src/malli/clj_kondo.cljc (hunk 1 above)
Comment (dvingo) [reply src/malli/clj_kondo.cljc:192]: This will allow you to emit the kondo config from a ClojureScript REPL because CLJS will need to call the Clojure code in this namespace from a macro.

**Comment by ikitommi** [comment]
Awesome, thanks!!

File: src/malli/core.cljc [hunk 3]
```diff
@@ -2387,6 +2387,9 @@
      (let [name' `'~(symbol (str name))
            ns' `'~(symbol (str *ns*))
            sym `'~(symbol (str *ns*) (str name))]
+       ;; in cljs we need to register the schema in clojure (the cljs compiler) so it is visible in the -function-schemas* map at macroexpansion time.
+       (when (some? (:ns &env))
+         (-register-function-schema! (symbol (str *ns*)) name value (meta name)))
```
Comment (ikitommi) [inline src/malli/core.cljc:2392]: Because of this change, this doesn't work now:

```clj
(def small-int [:int {:max 6}])

(defn plus [x] (inc x))
(m/=> plus [:=> [:cat :int] small-int])
```

... as the `small-int` is not resolved at macro-expansion time and the `small-int` is not resolved yet.

any ideas?

File: src/malli/instrument/cljs.clj [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+;;
+;; instrument
+;;
+
+(defn emit-instrument-fn [{:keys [gen filters] :as instrument-opts} {:keys [schema] :as schema-map} ns-sym fn-sym]
+  ;; gen is a function
+  (let [schema-map (assoc (select-keys schema-map [:gen :scope :report])
+                     ;; At macroexpansion time the schema will be a JVM object, this converts it to a JS object.
+                     :schema `(m/function-schema ~(m/form schema)))
```
Comment (ikitommi) [inline src/malli/instrument/cljs.clj:15]: these too happen at macroexpansion time, so will fail with local variables.

File: src/malli/instrument/cljs.clj [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  ([opts] (instrument* opts)))
+
+;;
+;; unstrument
+;;
+
+(defn emit-unstrument-fn [{:keys [schema filters] :as opts} ns-sym fn-sym]
+  (let [opts (assoc (select-keys opts [:gen :scope :report])
+               ;; At macroexpansion time the schema will be a JVM object, this converts it to a JS object.
+               :schema `(m/function-schema ~(m/form schema)))
```
Comment (ikitommi) [inline src/malli/instrument/cljs.clj:59]: here too.

Comment (ikitommi) [reply src/malli/instrument/cljs.clj:59]: calling `m/form` also removes the possibility to use non-registered Schema Instances:

```clj
(def MyInt (m/-simple-schema {:type 'MyInt, :pred int?}))

(m/validate MyInt 2) ; => true

(m/form MyInt) ; => MyInt

;; there is no way to construct the local instance from form.
(m/schema (m/form MyInt))
; =throws=> :malli.core/invalid-schema {:schema MyInt}
```

I'm testing if we can just use `schema` in the three places.

Comment (ikitommi) [reply src/malli/instrument/cljs.clj:59]: with my skills, we can't.

Comment (dvingo) [reply src/malli/instrument/cljs.clj:59]: ahh that's unfortunate! I will use these examples to see if I can get it working. I _think_ moving everything to happen at runtime (in cljs) should be possible. I will report back when I have an update.

Comment (ikitommi) [reply src/malli/instrument/cljs.clj:59]: awesome, looking forward to it!

**Comment by dvingo** [comment]
@kovasap  you're totally correct - this wasn't fully complete! 

I just pushed a PR to fix it up https://github.com/metosin/malli/pull/655

thanks for reporting that.

**Comment by kovasap** [comment]
Thanks!  And with this it seems like the documentation at https://github.com/metosin/malli/blob/master/docs/function-schemas.md#development-instrumentation is sufficient.

**Comment by dvingo** [comment]
I was just looking at that - the only piece that doesn't apply is
> and starts watching the function registry for changes. 

but the doc string for `start!` in dev.cljs calls this out.
